server/src/main/java/io/crate/replication/logical/MetadataTracker.java

```diff
@@ -201,36 +209,97 @@ static ClusterState updateMetadata(String subscriptionName, ClusterState subscri
+                isIndexMetadataUpdated =
+                    updateIndexMetadataSettings(updatedIndexMetadataBuilder, publisherIndexMetadata, subscriberIndexMetadata, indexScopedSettings) ||
+                    updateIndexMetadataMapping(updatedIndexMetadataBuilder, publisherIndexMetadata, subscriberIndexMetadata);
+
+                if (isIndexMetadataUpdated) {
+                    updatedMetadataBuilder.put(updatedIndexMetadataBuilder.build(), true);
                 }
+                updateClusterState = updateClusterState || isIndexMetadataUpdated;
             }
         }
-        if (isUpdated) {
+        if (updateClusterState) {
             if (LOGGER.isDebugEnabled()) {
                 LOGGER.debug("Updated index metadata for subscription {}", subscriptionName);
             }
-            return ClusterState.builder(subscriberClusterState).metadata(metadataBuilder).build();
+            return ClusterState.builder(subscriberClusterState).metadata(updatedMetadataBuilder).build();
         } else {
             return subscriberClusterState;
         }
     }
 
+    private static boolean updateIndexMetadataMapping(IndexMetadata.Builder updatedIndexMetadata,
+                                              IndexMetadata publisherIndexMetadata,
+                                              IndexMetadata subscriberIndexMetadata) {
+        var publisherMapping = publisherIndexMetadata.mapping();
+        var subscriberMapping = subscriberIndexMetadata.mapping();
+        if (publisherMapping != null && subscriberMapping != null) {
+            if (publisherIndexMetadata.getMappingVersion() > subscriberIndexMetadata.getMappingVersion()) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Updated index mapping {} for subscription {}", subscriberIndexMetadata.getIndex().getName(), publisherMapping.toString());
+                }
+                updatedIndexMetadata.putMapping(publisherMapping).mappingVersion(publisherIndexMetadata.getMappingVersion());
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static boolean updateIndexMetadataSettings(IndexMetadata.Builder updatedIndexMetadata,
+                                               IndexMetadata publisherMetadata,
+                                               IndexMetadata subscriberMetadata,
+                                               IndexScopedSettings indexScopedSettings) {
+        var isUpdated = false;
+        var publisherMetadataSetting = publisherMetadata.getSettings();
+        var subscriberMetadataSetting = subscriberMetadata.getSettings();
+        var newSubscriberIndexMetadataSettings = Settings.builder();
+        // Take all initial settings from the subscriber
+        for (var key : subscriberMetadataSetting.keySet()) {
+            newSubscriberIndexMetadataSettings.copy(key, subscriberMetadataSetting);
+        }
```

Comment:
Any reason not to use `newSubscriberIndexMetadataSettings.put(subscriberMetadataSetting)`?

Reply:
Sorry maybe I wasn't clear, I'm talking about the `Settings.Builder.put(Settings settings)` API.

Reply:
Got it, thanks.